### PR TITLE
Auth

### PR DESCRIPTION
**Pull request by maxbates**
(Max: Creating a pull request so I can create some general comments on it)

### CODE REVIEW

File: server/utils/local-auth.js [hunk 1]
```diff
@@ -0,0 +1,40 @@
+import express from 'express';
+const router = express.Router();
+
+var defaultUser = {
```
Comment (maxbates) [inline server/utils/local-auth.js]: lint things... prefer `const` etc.

File: src/components/authentication/account.js [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+  get passwordConfirm() {
+    return this.refs.passwordConfirm.value.trim();
+  }
+
+  /**
+   * display server errors in the most logical way
+   */
+  showServerErrors(json) {
+    invariant(json && json.message, 'We expected an error message');
+    if (json.path === 'email' && json.type === 'unique violation') {
```
Comment (maxbates) [inline src/components/authentication/account.js]: i feel like.... you should use variables not string checks. and why not a switch?

Comment (maxbates) [inline src/components/authentication/account.js]: Actually, really, why not create a react element which creates the input, does the validation, accepts the error message to show...

File: src/components/authentication/account.js [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+      payload.lastName = this.lastName;
+    }
+    if (this.emailAddress) {
+      payload.email = this.emailAddress;
+    }
+    if (this.password) {
+      payload.newPassword = this.password;
+    }
+
+    fetch(endPoint, {
```
Comment (maxbates) [inline src/components/authentication/account.js]: This should go into a new file `middleware/auth.js` or something

File: src/components/authentication/forgot.js [hunk 4]
```diff
@@ -1,18 +1,105 @@
...
+    // reset errors
+    this.setState({
+      emailError: false,
+      text: '&nbsp;',
+    });
+
+    // get the API end point
+    const endPoint = `${window.location.origin}/auth/forgot-password`;
+
+    fetch(endPoint, {
```
Comment (maxbates) [inline src/components/authentication/forgot.js]: This should go into a new file `middleware/auth.js` or something

File: src/components/authentication/reset.js [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+    // submission occurs via REST not form submission
+    evt.preventDefault();
+    // client side validation first
+    if (this.clientValidation()) {
+      return;
+    }
+    // get the API end point
+    const endPoint = `${window.location.origin}/auth/reset-password`;
+
+    fetch(endPoint, {
```
Comment (maxbates) [inline src/components/authentication/reset.js]: This should go into a new file `middleware/auth.js` or something

File: src/components/authentication/signin.js [hunk 6]
```diff
@@ -1,22 +1,133 @@
...
+  // on form submission, first perform client side validation then submit
+  // to the server if that goes well.
+  onSubmit(evt) {
+    // submission occurs via REST not form submission
+    evt.preventDefault();
+
+    // get the API end point
+    const endPoint = `${window.location.origin}/auth/login`;
+
+    fetch(endPoint, {
```
Comment (maxbates) [inline src/components/authentication/signin.js]: This should go into a new file `middleware/auth.js` or something

File: src/components/authentication/signup.js [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+    return this.refs.passwordConfirm.value.trim();
+  }
+  get tos() {
+    return this.refs.tos.checked;
+  }
+
+  /**
+   * display server errors in the most logical way
+   */
+  showServerErrors(json) {
```
Comment (maxbates) [inline src/components/authentication/signup.js]: brittle..

File: src/components/authentication/userwidget.js [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+    this.setState({
+      menuOpen: false,
+    });
+  }
+
+  signOut() {
+
+    const endPoint = `${window.location.origin}/auth/logout`;
+
+    fetch(endPoint, {
```
Comment (maxbates) [inline src/components/authentication/userwidget.js]: middleware module

File: src/components/ribbongrunt.js [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+// MS display time for grunt messages
+const DISPLAY_TIME = 5000;
+
+class RibbonGrunt extends Component {
+
+  static propTypes = {
+    gruntMessage: PropTypes.string,
+  };
+
+  constructor() {
```
Comment (maxbates) [inline src/components/ribbongrunt.js]: unnecessary

File: src/containers/GlobalNav.js [hunk 10]
```diff
@@ -40,14 +43,14 @@ class GlobalNav extends Component {
             {
               text: 'Recent Projects',
```
Comment (maxbates) [inline src/containers/GlobalNav.js]: you can get rid of this label if youre hiding the projects

File: src/containers/GlobalNav.js [hunk 11]
```diff
@@ -221,6 +224,27 @@ class GlobalNav extends Component {
             },
           ],
         },
+        {
+          text: 'AUTHENTICATION',
```
Comment (maxbates) [inline src/containers/GlobalNav.js]: make nice

File: src/containers/authentication/authenticationforms.js [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+                payload={form}
+                closeOnClickOutside
+                closeModal={(buttonText) => {
+                  this.props.uiShowAuthenticationForm('none')
+                }}
+              />;
+    }
+
+    return (
+      <div>
```
Comment (maxbates) [inline src/containers/authentication/authenticationforms.js]: div wrap unnecessary?

File: src/routes.js [hunk 13]
```diff
@@ -5,12 +5,22 @@ import App from './containers/App';
...
-import SignInPage from './components/authentication/SignInPage';
+import HomePage from './components/homepage';
 import SupportPage from './components/SupportPage';
 
 //Routes are specified as a separate component so they can hotloaded
 //see: https://github.com/rackt/redux-router/issues/44#issuecomment-140198502
 
+function requireAuth(nextState, replace) {
+  //if (!auth.loggedIn()) {
+  if (true) {
```
Comment (maxbates) [inline src/routes.js]: this is temporary?

File: src/styles/Modal.css [hunk 14]
```diff
@@ -29,10 +29,10 @@
 }
```
Comment (maxbates) [inline src/styles/Modal.css]: prefer nested properties

File: src/styles/authenticationforms.css [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+.authentication-form {
+  width: 500px;
+  padding-bottom: 1rem;
+  display: flex;
+  flex-direction: column;
+  align-items: center;
+  font-size: 1.5rem;
+}
+
+.authentication-form * {
```
Comment (maxbates) [inline src/styles/authenticationforms.css]: prefer nested properties

**Comment by maxbates** [comment]
Biggest things:
- create `middleware/auth.js` or something, rather than direct fetch requests
- The auth forms are all pretty clunky.. why not create a react element to delegate a lot of this handling to? 
- UI forms are kinda brittle - use variables at least for strings, not string equality checks
- more tests!

And then some lint / consistency things:

- prefer one property per line in react elements
- prefer functions as instance properties rather than on class prototype:
```
// :(
class Sadness {
  myProperty() => { ... }
  <div onClick={this.myProperty.bind(this)}></div>
}
// :)
class Yay {
  myProperty = (evt) => { ... }
  <div onClick={this.myProperty}></div>
}
```
- similarly for state, rather than in constructor:
```
class Sadness {
  constructor() {
    this.state = {property: value};
  }
}
class Yay {
  state = {property: value};
}
```
- CSS should use nesting and the `&` extension, like SCSS
- put variables in the variables sheet, unless it really makes sense not to... but things like colors, those should be made consistent. `variables.css` should serve same sort of utility / flexibility as it does in bootstrap

**Comment by maxbates** [comment]
NPM script names should change to something like (`auth` -> `run` and `run` -> `local`) so that the auth version is the default one

**Comment by dhylbert** [comment]
@maxbates you want the default scripts to require access to git.autodesk.com?

**Comment by maxbates** [comment]
@duncanmeech you are committing to maintaining these dialog boxes I hope! If you want others to maybe do that later then you should use variables especially for your validation checks + react components to simplify

**Comment by maxbates** [comment]
@duncanmeech login and logout should be actions. Your components are still handling setting state - the actions should do that. Components shouldnt be responsible for all that state management (e.g. userwidget)

**Comment by maxbates** [comment]
@duncanmeech ribbongrunt should be child of globalNav

File: src/middleware/api.js [hunk 16]
```diff
@@ -24,55 +24,117 @@ export const searchPath = (id) => serverRoot + 'search/' + id;
...
+      throw err;
+    });
+};
+
+export const forgot = (email) => {
+  const body = { email }
+  const stringified = JSON.stringify(body);
+  return fetch(serverRoot + `auth/forgot-password`, headersPost(stringified))
+    .then(resp => resp.json())
+    .then(json => {
```
Comment (maxbates) [inline src/middleware/api.js]: pls remove

File: src/middleware/api.js [hunk 17]
```diff
@@ -24,55 +24,117 @@ export const searchPath = (id) => serverRoot + 'search/' + id;
...
+    });
+};
+
+export const reset = (email, forgotPasswordHash, newPassword) => {
+  const body = { email, forgotPasswordHash, newPassword }
+  const stringified = JSON.stringify(body);
+  return fetch(serverRoot + `auth/reset-password`, headersPost(stringified))
+    .then(resp => resp.json())
+    .then(json => {
+      return json;
```
Comment (maxbates) [inline src/middleware/api.js]: remove

File: src/middleware/api.js [hunk 18]
```diff
@@ -24,55 +24,117 @@ export const searchPath = (id) => serverRoot + 'search/' + id;
...
+export const updateAccount = (payload) => {
+  const body = payload;
+  const stringified = JSON.stringify(body);
+
+  return fetch(serverRoot + `auth/update-all`, headersPost(stringified))
     .then(resp => resp.json())
     .then(json => {
-      sessionKey = json.sessionkey;
-      return sessionKey;
+      return json;
```
Comment (maxbates) [inline src/middleware/api.js]: remove

File: src/middleware/api.js [hunk 19]
```diff
@@ -24,55 +24,117 @@ export const searchPath = (id) => serverRoot + 'search/' + id;
...
+  return fetch(serverRoot + `auth/logout`, headersGet());
+};
+
+// use established sessionKey to get the user object
+export const getUser = () => {
+  return fetch(serverRoot + `auth/current-user`, headersGet())
+    .then(resp => {
+      return resp.json();
+    })
+    .catch((err) => {
```
Comment (maxbates) [inline src/middleware/api.js]: remove

File: src/routes.js [hunk 20]
```diff
@@ -14,20 +14,19 @@ import SupportPage from './components/SupportPage';
...
-
+    <Route component={RouteWrapper}>
+      <Route path="/homepage/account" component={HomePage}/>
+      <Route path="/project/:projectId" component={ProjectPage}/>
+    </Route>
     <Route path="/about" component={AboutPage}/>
-    <Route path="/signin" component={SignInPage}/>
     <Route path="/support" component={SupportPage}/>
-    <Route path="/project/:projectId" component={ProjectPage}/>
+    <Route path="/homepage/signin" component={HomePage}/>
```
Comment (maxbates) [inline src/routes.js]: Are these routes necessary?

File: src/middleware/api.js (hunk 19 above)
Comment (dhylbert) [inline src/middleware/api.js]: Aren't tests dependent on these methods?

Comment (maxbates) [inline src/middleware/api.js]: just the final then / catch parts

**Pull request merged**